### PR TITLE
metadata: extract metadata from appstream

### LINKED ISSUE #1693: Develop metadata handler system
Estimate: 1 week

Currently Snapcraft requires all its metadata (snap name, version, etc.) to be specified in the snapcraft.yaml. However, lots of parts already contain such information, whether from a setup.py, XML, AppStream, etc.

I’m considering this task to only be for existing metadata (name, version, summary, etc.). Adding support for an AppStream ID etc. is a separate task

### PR DESCRIPTION
**Pull request by elopio**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----
See the design discussion in the [forum](https://forum.snapcraft.io/t/extracting-existing-data-from-projects-to-feed-into-snap-yaml/2285/8). Closes #1693

### CODE REVIEW

**Comment by sergiusens** [comment]
I was thinking more that this should have been handled during prime by the pluginhandler storing the _values_ into the state tracker for the part, pushing them onto a stack; then during meta creation, this would be retrievable from this stack.

I say stack it instead of replace to get an idea of what is going on when weird things happen. Doing this in the pluginhandler will also ease the implementation of the helpers such as `set-description` and `set-name`

**Comment by sergiusens** [comment]
This is missing icon extraction and possibly desktop

**Comment by kyrofa** [comment]
I assume this is referencing the design [here](https://forum.snapcraft.io/t/extracting-existing-data-from-projects-to-feed-into-snap-yaml/2285/8)?

File: schema/snapcraft.yaml [hunk 1]
```diff
@@ -361,16 +364,20 @@ properties:
...
   slots:
     type: object
 required:
   - name
   - version
-  - summary
-  - description
   - parts
+# required by snap.yaml, but that can be filled later by other metadata
+# sources, like appstream.
```
Comment (kyrofa) [inline schema/snapcraft.yaml]: Excellent comment, thank you!

File: schema/snapcraft.yaml [hunk 2]
```diff
@@ -361,16 +364,20 @@ properties:
           prepare:
             type: string
             default: ''
+          parse-info:
+            type: array
```
Comment (kyrofa) [inline schema/snapcraft.yaml]: The array items are always going to be strings, right? We should probably flesh this schema definition out to enforce that.

**Comment by kyrofa** [comment]
Also, I thought the underlying metadata extraction was going to be its own PR, separate from the appstream implementation of it?

File: snapcraft/internal/meta/_appstream.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    ParseError
+)
+
+
+class AppstreamInfoParser:
+
+    def __init__(self, path: str) -> None:
+        self._path = path
+
+    def extract(self, keys: list) -> dict:
```
Comment (kyrofa) [inline snapcraft/internal/meta/_appstream.py]: I didn't realize this was an option, haha! You might consider using `Dict` though, it gives you the ability to specify key/value types as well.

File: snapcraft/internal/meta/_errors.py [hunk 4]
```diff
@@ -21,3 +21,72 @@ class CommandError(errors.SnapcraftError):
 
     def __init__(self, message: str) -> None:
         self.fmt = message
+
+
+class SnapMetaGenerationError(errors.SnapcraftError):
+    pass
+
+
+class SnapcraftYamlMissingRequiredValue(SnapMetaGenerationError):
```
Comment (kyrofa) [inline snapcraft/internal/meta/_errors.py]: Just looking at the errors, I can't figure out the difference between this error class and `MissingSnapcraftYamlKeys`.

File: snapcraft/internal/meta/_from_sources.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import os
+
+from snapcraft.internal.meta import (
+    _appstream,
+    _errors
+)
+
+
+def get_info(keys: list, config_data: dict, parts_dir: str) -> dict:
```
Comment (kyrofa) [inline snapcraft/internal/meta/_from_sources.py]: Consider using something more specific than `dict` here as well.

File: snapcraft/internal/meta/_errors.py (hunk 4 above)
Comment (kyrofa) [reply snapcraft/internal/meta/_errors.py]: After reading where it's used, it sounds like this should be part-specific, no?

File: snapcraft/internal/meta/_from_sources.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            if 'parse-info' in config_data['parts'][part]:
+                part_dir = os.path.join(parts_dir, part)
+                info_paths = config_data['parts'][part]['parse-info']
+                return _extract_info_from_part(
+                    keys,
+                    config_data['parts'][part]['plugin'],
+                    part_dir,
+                    info_paths)
+            else:
+                raise _errors.SnapcraftYamlMissingRequiredValue('parse-info')
```
Comment (kyrofa) [inline snapcraft/internal/meta/_from_sources.py]: Shouldn't this include the part name?

File: snapcraft/internal/meta/_errors.py [hunk 7]
```diff
@@ -21,3 +21,72 @@ class CommandError(errors.SnapcraftError):
...
+    fmt = (
+        "Failed to generate snap metadata: "
+        "Missing {key!r} from the 'snapcraft.yaml' file."
+    )
+
+    def __init__(self, key: str) -> None:
+        super().__init__(key=key)
+
+
+class WrongAdoptInfo(SnapMetaGenerationError):
```
Comment (kyrofa) [inline snapcraft/internal/meta/_errors.py]: How about `AdoptedPartMissingError`?

File: snapcraft/internal/meta/_from_sources.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        raise _errors.SnapcraftYamlMissingRequiredValue('adopt-info')
+
+
+def _extract_info_from_part(keys, plugin, part_dir, info_paths):
+    info = {}
+    for path in info_paths:
+        if keys:
+            source_full_paths = [
+                os.path.join(part_dir, step_dir, path)
+                for step_dir in ('install', 'build')]
```
Comment (kyrofa) [inline snapcraft/internal/meta/_from_sources.py]: What about parts that build out-of-source (e.g. the cmake plugin)? We should probably also search the `src/` dir, no?

Comment (sergiusens) [reply snapcraft/internal/meta/_from_sources.py]: We should revisit that, maybe it is fine to use src and not just for cmake; in most cases appstream.xml files in sources are found as `.in` templates to be finalized during build. Let's add it, it doesn't hurt and it a fallback mechanism anyways.

**Comment by sergiusens** [comment]
@kyrofa you are right, this needs to be split to solve #1693

**Comment by sergiusens** [comment]
@kyrofa and  #1694

**Pull request closed without merging**